casa/IO/MultiFile.cc

```diff
@@ -36,191 +37,752 @@
+    aio << itsInfo;
     aio.putend();
+    // Write the used and free blocknrs in packed format but outside
+    // the AipsIO object because that is limited to 4 GB.
+    for (const MultiFileInfo& fileInfo : info()) {
+      writeVector (cio, packIndex(fileInfo.blockNrs));
+    }
+    writeVector (cio, packIndex(freeBlocks()));
+    writeVector (cio, itsCRC);
+    // Calculate the size including the continuation blocknrs and number of
+    // actually used blocknrs.
+    // If continuation is needed, they might change and cannot
+    // be written yet.
     Int64 todo = mio.length();
+    DebugAssert (todo <= mio.allocated(), AipsError);
+    Int64 totalSize = todo + 2*sizeof(uInt) + (2 + itsHdrCont[0].blockNrs.size() +
+                              itsHdrCont[1].blockNrs.size()) * sizeof(Int64);
+    // Allocate a temp buffer if header too large or if O_DIRECT.
+    Bool hasRemainder = (totalSize > itsBlockSize);
+    MultiFileBuffer mfbuf(itsUseODirect || hasRemainder  ?  itsBlockSize:0,
+                          itsUseODirect);
+    char* iobuf = mfbuf.data();
+    // First write the remainder and adjust the header as needed.
+    if (hasRemainder) {
+      writeRemainder (mio, cio, iobuf);
+    } else {
+      // No remainder, so write the continuation blocknrs here.
+      // None are used in the current set.
+      writeVector (cio, itsHdrCont[0].blockNrs);
+      writeVector (cio, itsHdrCont[1].blockNrs);
+      itsNrContUsed[itsHdrContInx] = 0;
+      cio.write (2, itsNrContUsed);
+    }
+    // Writing the first part of the header is done as the last step.
     uChar* buf = const_cast<uChar*>(mio.getBuffer());
-    CanonicalConversion::fromLocal (buf, todo);       // header size
-    // Write the first part of the buffer at the beginning of the file.
-    // Use an aligned buffer for possible O_DIRECT alignment.
-    itsIO.seek (0);
+    todo = mio.length();
+    CanonicalConversion::fromLocal (buf+24, todo);  // header size
+    if (itsUseCRC) {
+      uInt crc = calcCRC (buf, todo);
+      crc = calcCRC (buf, todo);
+      CanonicalConversion::fromLocal (buf+20, crc);
+    }
+    if (itsTestMode > 0  &&  itsNrBlock > itsTestMode) {
+      // Mimic an abort by not writing the first header block.
+      cout << "MultiFile::writeHeader ended prematurely for nrBlock="
+           << itsNrBlock << endl;
+      itsTestMode = -2;    // indicate no more writing
+      return False;
+    }
     if (itsUseODirect) {
-      MultiFileBuffer mfbuf(itsBlockSize, itsUseODirect);
-      memcpy (mfbuf.data, buf, itsBlockSize);
-      itsIO.pwrite (itsBlockSize, 0, mfbuf.data);
+      memcpy (iobuf, buf, itsBlockSize);
+      itsIO->pwrite (itsBlockSize, 0, iobuf);
     } else {
-      itsIO.pwrite (itsBlockSize, 0, buf);
+      itsIO->pwrite (itsBlockSize, 0, buf);
+    }
+    return True;
+  }
+
+  void MultiFile::writeRemainder (MemoryIO& mio, CanonicalIO& cio,
+                                  char* iobuf)
```

Comment:
What are the requirements of this pointer?

Reply:
This is indeed somewhat unclear. I've changed the interface to using a MultiFileBuffer.